numpy/lib/function_base.py

```diff
@@ -3637,80 +3875,157 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
     if axis is None:
         axis = 0
 
-    Nx = ap.shape[axis]
-    indices = q * (Nx - 1)
+    # move axis to -1 for np.vectorize() operations.  Will move back later.
+    ap = np.swapaxes(ap, axis, -1)
+
+    if weights is None:
+        Nx = ap.shape[-1]
+        indices = q * (Nx - 1)
+    else:
+        # need a copy of weights for later array assignment
+        weights = np.swapaxes(weights.astype('f8'), axis, -1)
+        weights[weights < 0.] = 0.  # negative weights are treated as 0
+        # values with weight=0 are assigned minimum value and later moved left
+        abs_min = np.amin(ap)
+        ap[weights == 0.] = abs_min - 1.
+
+        def _sort_by_index(vector, vec_indices):
+            return vector[vec_indices]
+        # this func vectorizes sort along axis
+        arraysort = np.vectorize(_sort_by_index, signature='(i),(i)->(i)')
+
+        ind_sorted = np.argsort(ap, axis=-1)  # sort values long axis
+        ap_sorted = arraysort(ap, ind_sorted).astype('f8')
+
+        n = np.isnan(ap_sorted[..., -1:])
+        if n.ndim > 1:
+            n = np.swapaxes(n, axis, -1)
+
+        ws_sorted = arraysort(weights, ind_sorted).astype('f8')
+        ws_sorted[np.isnan(ap_sorted)] = 0.  # neglect nans from calculation
+        nonzero_w_inds = ws_sorted > 0.
+
+        cum_w = ws_sorted.cumsum(axis=-1)
+        cum_w_max = cum_w.max(axis=-1)
```

Comment:
Do you want a `keepdims` here?

Reply:
hmm I don't think so.  I'm not sure how that would help me given how everything else is set up.

Reply:
It would eliminate the need for the transpose